File: sample_algorithms/Sorting/bubblesort/javascript/bbsort.js
import Tracer from './runtime/tracer.js';
import { swap } from './helpers.js';

function bubbleSort(arr, tracer) {
    const n = arr.length;
    
    for (let i = 0; i < n; i++) {
        let swapped = false;  
        for (let j = 0; j < n - i - 1; j++) {
            tracer.add_state([...arr], { 
                comparing: [j, j + 1],
                indexVars: ['i', 'j'],
                variables: { i, j } 
            });
            if (arr[j] > arr[j + 1]) {
                swap(arr, j, j + 1);
                tracer.add_state([...arr], { 
                    swapped: [j, j + 1],
                    indexVars: ['i', 'j'],
                    variables: { i, j } 
                });
                swapped = true;
            }
        }
        
        if (!swapped) {
            break;
        }
    }
    tracer.add_state([...arr]); // Complete state
    return [arr, tracer];
}

// [PARAMS]
const originalArr = [64, 34, 25, 12, 22, 11, 90];
// [/PARAMS]
const tracer = new Tracer('sorting'); // Tracer instance
const [sortedArr] = bubbleSort([...originalArr], tracer);

console.log('Original array:', originalArr);
console.log('Sorted array:', sortedArr);

// Output tracer data for visualization
tracer.finalize();

File: sample_algorithms/Sorting/bubblesort/python/bbsort.py
from runtime.tracer import Tracer
from helpers import swap

# [ALGORITHM]
def bubble_sort(arr, tracer):
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            # Show comparison (yellow)
            tracer.add_state(
                arr.copy(), 
                comparing=[j, j + 1],
                indexVars=['i', 'j'],
                variables={'i': i, 'j': j}
            )
            if arr[j] > arr[j + 1]:
                swap(arr, j, j + 1)
                # Show swap result (green)
                tracer.add_state(
                    arr.copy(), 
                    swapped=[j, j + 1],
                    indexVars=['i', 'j'],
                    variables={'i': i, 'j': j}
                )
                swapped = True
        if not swapped:
            break
    tracer.add_state(arr.copy()) #Complete state
    return arr, tracer

# [TEST]
if __name__ == "__main__":
    # [PARAMS]
    original_arr = [92, 14, 461, 1122, 235, 9, 127]
    # [/PARAMS]
    sorted_arr, tracer = bubble_sort(original_arr.copy(), Tracer(category='sorting'))

    print(f'Original array: {original_arr}')
    print(f'Sorted array: {sorted_arr}')
    
    tracer.finalize()